{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seed: 1763811343\n",
      "IQ-TREE version 2.3.5.lib COVID-edition for Linux 64-bit built Aug  8 2024\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams, Ly Trong Nhan.\n",
      "\n",
      "Host:    9171405acd0c (SSE4.2, 11 GB RAM)\n",
      "Time:    Mon Aug 12 04:05:30 2024\n",
      "Kernel:  SSE2\n",
      "\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "WARNING: 162 sites contain only gaps or ambiguous characters.\n",
      "Alignment has 4 sequences with 3009 columns, 87 distinct patterns\n",
      "38 parsimony-informative, 725 singleton sites, 2246 constant sites\n",
      "            Gap/Ambiguity  Composition  p-value\n",
      "Analyzing sequences: done in 0.000118017 secs using 92.36% CPU\n",
      "   1  Human         5.68%    passed     29.79%\n",
      "   2  Chimpanzee    5.68%    passed     39.39%\n",
      "   3  Rhesus        5.78%    passed     15.40%\n",
      "   4  Mouse         9.07%    failed      0.00%\n",
      "****  TOTAL         6.56%  1 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "enter initializeParams\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.003 seconds\n",
      "\n",
      "NOTE: 0 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -6872.908\n",
      "2. Current log-likelihood: -6794.267\n",
      "3. Current log-likelihood: -6739.418\n",
      "4. Current log-likelihood: -6736.949\n",
      "Optimal log-likelihood: -6736.946\n",
      "Rate parameters:  A-C: 1.00000  A-G: 3.82025  A-T: 1.00000  C-G: 1.00000  C-T: 3.82025  G-T: 1.00000\n",
      "Base frequencies:  A: 0.363  C: 0.185  G: 0.217  T: 0.234\n",
      "Parameters optimization took 4 rounds (0.007 sec)\n",
      "Wrote distance file to... \n",
      "Computing ML distances based on estimated model parameters...\n",
      "Calculating distance matrix: done in 0.000669956 secs using 98.36% CPU\n",
      "Computing ML distances took 0.001257 sec (of wall-clock time) 0.001256 sec (of CPU time)\n",
      "Setting up auxiliary I and S matrices: done in 0.00019002 secs using 94.2% CPU\n",
      "Computing RapidNJ tree took 0.001323 sec (of wall-clock time) 0.001287 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -6736.946\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 99 parsimony trees... 0.011 second\n",
      "Computing log-likelihood of 2 initial trees ... 0.000 seconds\n",
      "Current best score: -6736.946\n",
      "\n",
      "Do NNI search on 3 best initial trees\n",
      "Finish initializing candidate tree set (3)\n",
      "Current best tree score: -6736.946 / CPU time: 0.014\n",
      "Number of iterations: 3\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "UPDATE BEST LOG-LIKELIHOOD: -6736.946\n",
      "Iteration 10 / LogL: -6737.087 / Time: 0h:0m:0s\n",
      "UPDATE BEST LOG-LIKELIHOOD: -6736.946\n",
      "Iteration 20 / LogL: -6757.776 / Time: 0h:0m:0s\n",
      "Iteration 30 / LogL: -6757.568 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 40 / LogL: -6736.946 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 50 / LogL: -6737.087 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 60 / LogL: -6737.072 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 70 / LogL: -6737.046 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 80 / LogL: -6737.072 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 90 / LogL: -6736.946 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 100 / LogL: -6737.072 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "TREE SEARCH COMPLETED AFTER 101 ITERATIONS / Time: 0h:0m:0s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -6736.946\n",
      "Optimal log-likelihood: -6736.946\n",
      "Rate parameters:  A-C: 1.00000  A-G: 3.82025  A-T: 1.00000  C-G: 1.00000  C-T: 3.82025  G-T: 1.00000\n",
      "Base frequencies:  A: 0.363  C: 0.185  G: 0.217  T: 0.234\n",
      "Parameters optimization took 1 rounds (0.000 sec)\n",
      "BEST SCORE FOUND : -6736.946\n",
      "Total tree length: 0.353\n",
      "\n",
      "Total number of iterations: 101\n",
      "CPU time used for tree search: 0.028 sec (0h:0m:0s)\n",
      "Wall-clock time used for tree search: 0.030 sec (0h:0m:0s)\n",
      "Total CPU time used: 0.062 sec (0h:0m:0s)\n",
      "Total wall-clock time used: 0.074 sec (0h:0m:0s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                build_tree_1763811343.iqtree\n",
      "  Maximum-likelihood tree:       build_tree_1763811343.treefile\n",
      "  Likelihood distances:          build_tree_1763811343.mldist\n",
      "  Screen log file:               build_tree_1763811343.log\n",
      "\n",
      "Date and Time: Mon Aug 12 04:05:30 2024\n"
     ]
    }
   ],
   "source": [
    "# parse yaml\n",
    "from cogent3 import load_aligned_seqs\n",
    "from piqtree2 import build_tree\n",
    "from piqtree2.model import DnaModel\n",
    "\n",
    "aln = load_aligned_seqs(\"../tests/data/example.fasta\").take_seqs([\"Human\", \"Chimpanzee\", \"Rhesus\", \"Mouse\"])\n",
    "\n",
    "tree = build_tree(aln, DnaModel.HKY)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'newick': '(Human,Chimpanzee,(Rhesus,Mouse)edge.0)',\n",
       " 'edge_attributes': {'Human': {'length': 0.0063211201,\n",
       "   'A/C': 1.0,\n",
       "   'A/G': 3.82025079,\n",
       "   'A/T': 1.0,\n",
       "   'C/G': 1.0,\n",
       "   'C/T': 3.82025079,\n",
       "   'G/T': 1.0},\n",
       "  'Chimpanzee': {'length': 0.002967578,\n",
       "   'A/C': 1.0,\n",
       "   'A/G': 3.82025079,\n",
       "   'A/T': 1.0,\n",
       "   'C/G': 1.0,\n",
       "   'C/T': 3.82025079,\n",
       "   'G/T': 1.0},\n",
       "  'Rhesus': {'length': 0.0228519739,\n",
       "   'A/C': 1.0,\n",
       "   'A/G': 3.82025079,\n",
       "   'A/T': 1.0,\n",
       "   'C/G': 1.0,\n",
       "   'C/T': 3.82025079,\n",
       "   'G/T': 1.0},\n",
       "  'Mouse': {'length': 0.3072009029,\n",
       "   'A/C': 1.0,\n",
       "   'A/G': 3.82025079,\n",
       "   'A/T': 1.0,\n",
       "   'C/G': 1.0,\n",
       "   'C/T': 3.82025079,\n",
       "   'G/T': 1.0},\n",
       "  'edge.0': {'length': 0.01373649616,\n",
       "   'A/C': 1.0,\n",
       "   'A/G': 3.82025079,\n",
       "   'A/T': 1.0,\n",
       "   'C/G': 1.0,\n",
       "   'C/T': 3.82025079,\n",
       "   'G/T': 1.0},\n",
       "  'root': {'length': None, 'lnL': -6736.94578464}},\n",
       " 'type': 'cogent3.core.tree.PhyloNode',\n",
       " 'version': '2024.5.7a1'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree.to_rich_dict()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "c312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
